[HC-1397] - Improve udev rules solutions (#595)
arduino/help-center-content: content/Software Support/Upload/Fix-udev-rules-on-Linux.md

If you encounter failed uploads, you can resolve the issue by installing the unified Arduino `udev` rules script.

---

## Related Error Messages

If your `udev` rules are missing or configured incorrectly, you may see errors similar to the following depending on your board:

| Boards | Related error output |
| :--- | :--- |
| **GIGA R1 WiFi / Portenta H7 / Portenta C33 / UNO R4** | `dfu-util: Cannot open DFU device 2341:0366 found on devnum X (LIBUSB_ERROR_ACCESS)`<br>`dfu-util: No DFU capable USB device available`<br>`Failed uploading: uploading error: exit status 74` |
| **Nicla Sense ME** | `Error: unable to open CMSIS-DAP device 0x2341:0x60`<br>`Error: unable to find a matching CMSIS-DAP device`<br>`Failed uploading: uploading error: exit status 1` |
| **Nano Every / UNO WiFi Rev2** | `avrdude: jtagmkII_getsync(): sign-on command: status -1`<br>`avrdude: usbdev_open(): cannot open device: Permission denied`<br>`Failed uploading: uploading error: exit status 1` |

> | **Nano RP2040 Connect** | `Failed uploading: uploading error: exit status 1` |

Reviewer: It'd be nice to include UNO Q in this table as well.

Author: Done

id: 9005041052444
---

On Linux, the `udev` subsystem manages access to connected USB devices. By default, standard user accounts do not have permission to write directly to serial, USB, or DFU (Device Firmware Upgrade) interfaces of connected hardware.

`udev` rules configure the system to recognize Arduino boards and grant your user account the necessary permissions to access, upload code, and debug. Missing or incorrect `udev` rules can result in failed uploads, as Arduino IDE and other development tools will not be able to access the board (especially when it resets to bootloader or DFU mode).

> If you encounter failed uploads, you can resolve the issue by installing the unified Arduino `udev` rules script.

Reviewer: ```suggestion
On Linux, `udev` rules configure the system to recognize Arduino boards and grant your user account the necessary permissions to access, upload code, and debug. Missing or incorrect `udev` rules can result in failed board detection or uploads, as Arduino IDE and other development tools will not be able to access the board.

You can resolve the issue by configuring the required rules using an [automated bash script](#automatic-configuration).
```

Suggestion for streamlined introduction. It also removed the "unified" word, which is no longer a concept relevant to the instructions.

Author: Done on last commit

---

## Related Error Messages

If your `udev` rules are missing or configured incorrectly, you may see errors similar to the following depending on your board:

| Boards | Related error output |
| :--- | :--- |
| **GIGA R1 WiFi / Portenta H7 / Portenta C33 / UNO R4** | `dfu-util: Cannot open DFU device 2341:0366 found on devnum X (LIBUSB_ERROR_ACCESS)`<br>`dfu-util: No DFU capable USB device available`<br>`Failed uploading: uploading error: exit status 74` |
| **Nicla Sense ME** | `Error: unable to open CMSIS-DAP device 0x2341:0x60`<br>`Error: unable to find a matching CMSIS-DAP device`<br>`Failed uploading: uploading error: exit status 1` |
| **Nano Every / UNO WiFi Rev2** | `avrdude: jtagmkII_getsync(): sign-on command: status -1`<br>`avrdude: usbdev_open(): cannot open device: Permission denied`<br>`Failed uploading: uploading error: exit status 1` |
| **Nano RP2040 Connect** | `Failed uploading: uploading error: exit status 1` |
| **UNO Q** | `Error: exit status 1: adb: insufficient permissions for device: missing udev rules? user is in the plugdev group`<br>`Failed uploading: uploading error: exit status 1` |

---

## How to Install the Unified udev Rules Script

> We have created a unified script that configures `udev` rules for all supported Arduino boards in one step.

Reviewer: ```suggestion
## Automatic configuration

You can configure rules for all Arduino boards by downloading and running `arduino-udev-setup.sh`.
```

Since we are no longer using "unified" in the script name.

Author: Done

  ```bash
  wget https://content.arduino.cc/assets/arduino-udev-setup.sh
  ```

  *or*

  ```bash
  curl -O https://content.arduino.cc/assets/arduino-udev-setup.sh
  ```

### Step 2: Open the terminal and navigate to the file

If you downloaded the file via Web Browser:

Open your Terminal and navigate to your downloads folder manually:

  ```bash
  cd ~/Downloads
  ```

### Step 3: Run the script

You can run the script directly without modifying permissions by using bash:

```bash
sudo bash arduino-udev-setup.sh
```

When prompted, enter your Linux user password and press Enter.

#### Optional: Make the script executable

If you prefer, you can make the script executable and run it directly:

```bash
chmod +x arduino-udev-setup.sh
sudo ./arduino-udev-setup.sh
```

### Step 4: Run the script as root

The script needs to create a configuration file in a system directory (`/etc/udev/rules.d/`) and reload the `udev` service, which requires administrator (root) privileges.

Run the script using `sudo`:

```bash
sudo ./arduino-udev-setup.sh
```

When prompted, enter your Linux user password and press **Enter**.

> ### Step 5: Reconnect your board

Reviewer: ~~~suggestion
### Step 3: Run the script as root

The script needs to create a configuration file in a system directory (`/etc/udev/rules.d/`) and reload the `udev` service, which requires administrator (root) privileges.

Run the script using bash:

```bash
sudo bash arduino-udev-setup.sh
```

When prompted, enter your Linux user password and press **Enter**.

#### Step 4: Reconnect your board
~~~

## Step 3 and Step 4 are redundant

- Step 3 tells the user to run sudo bash `arduino-udev-setup.sh` and enter their password.
- Step 4 tells the user why it needs to run as root, and then tells them to run `sudo ./arduino-udev-setup.sh` and enter their password again.

## Simplify script execution 

Based on the instructions, `sudo bash arduino-udev-setup.sh` is the superior choice because it bypasses the need for the user to modify file permissions. Multiple ways to do the exact same thing increase cognitive load. I suggest removing the other option.

Author: Done on last commit

Once the script completes successfully:

1. Unplug your Arduino board from the USB port.
2. Plug the board back into the USB port.

> This forces the system to reload and apply the new `udev` rules to your device. Try uploading your sketch again in the Arduino IDE!

Reviewer: ```suggestion
This forces the system to reload and apply the new `udev` rules to your device. Try uploading your sketch again in Arduino IDE!
```

Arduino IDE is a proper noun, so "the" can be omitted.

| **Nicla Sense ME** | `Error: unable to open CMSIS-DAP device 0x2341:0x60`<br>`Error: unable to find a matching CMSIS-DAP device`<br>`Failed uploading: uploading error: exit status 1` |
| **Nano Every / UNO WiFi Rev2** | `avrdude: jtagmkII_getsync(): sign-on command: status -1`<br>`avrdude: usbdev_open(): cannot open device: Permission denied`<br>`Failed uploading: uploading error: exit status 1` |
| **Nano RP2040 Connect** | `Failed uploading: uploading error: exit status 1` |

---

## How to Install the Unified udev Rules Script

We have created a unified script that configures `udev` rules for all supported Arduino boards in one step.

Follow these steps to download and run the script:

### Step 1: Download the script

You can download the script either using your web browser or directly in the terminal:

* **Option A (Web Browser):** Go to the [unified-arduino-udev-rules.sh file on GitHub](https://github.com/arduino/help-center-content/blob/main/utilities/unified-arduino-udev-rules.sh). Click the **Download raw file** button (the download icon on the top right) to save the file to your computer.

>   ![The "Download raw file" button on github.com](img/github-udev-download.png)

Reviewer 2: You can just use the raw link directly 🙂 (also: is the next line a cut and paste issue?)

```suggestion
* **Option A (Web Browser):** Download the [unified-arduino-udev-rules.sh file on GitHub](https://github.com/arduino/help-center-content/raw/main/utilities/unified-arduino-udev-rules.sh).
```

Reviewer 2: IMO a direct download from the content website would be way better than having to navigate the Github UI to get a file.

I thought at least providing the direct raw link would be easier, but maybe it is more complicated for mobile users. I'm not one you should listen to on this though, I only use `wget` and `vi`... :rofl:

Reviewer: I think we can host it on content.arduino.cc

We'll have a look at that when the script is ready

Author: We will publish it via GitHub first, and then consider the option of host it on content.arduino.cc and update the article. Thank you for the feedback.

> * **Option B (Terminal):** Open your terminal and run one of the following command to download the file directly into your current directory:

Reviewer: ```suggestion
* **Option B (Terminal):** Open your terminal and run one of the following commands to download the file directly into your current directory:
```

  ```bash
  wget https://raw.githubusercontent.com/arduino/help-center-content/main/utilities/unified-arduino-udev-rules.sh
  ```

  *or*

  ```bash
  curl -O https://raw.githubusercontent.com/arduino/help-center-content/main/utilities/unified-arduino-udev-rules.sh
  ```

### Step 2: Open the terminal and navigate to the file

If you downloaded the file via Web Browser:

Open your Terminal and navigate to your downloads folder manually:

  ```bash
  cd ~/Downloads
  ```

### Step 3: Grant execution permissions

Before running the script, you must grant it execution permissions so the operating system knows it is allowed to run.

Run the following command:

```bash

> chmod +x unified-arduino-udev-rules.sh

Reviewer 3: You can also skip this part and run `bash unified-arduino-udev-rules.sh` at the next step to execute it.

Author: Suggestion applied on [this ](https://github.com/arduino/help-center-content/pull/595/changes/bbfded8b841fd07423351b17a9f8328644630f79)commit
